File: src/lib/video/scene-detector.ts
import ffmpeg from "fluent-ffmpeg";

export interface DetectedScene {
  startMs: number;
  endMs: number;
  score: number;
}

/**
 * Detects scene boundaries in a video using FFmpeg's scene detection filter.
 * Uses `select='gt(scene,THRESHOLD)'` with showinfo to find scene-change timestamps.
 *
 * @param filePath  Path to the video file
 * @param threshold Scene-change sensitivity (0–1). Lower = more scenes detected. Default 0.3.
 * @returns Array of detected scenes with start/end times and confidence scores.
 */
export function detectScenes(
  filePath: string,
  threshold: number = 0.3
): Promise<DetectedScene[]> {
  return new Promise((resolve, reject) => {
    const timestamps: { timeSec: number; score: number }[] = [];
    let durationMs = 0;

    ffmpeg.ffprobe(filePath, (probeErr, data) => {
      if (probeErr) {
        return reject(new Error(`ffprobe failed: ${probeErr.message}`));
      }

      durationMs = Math.round((data.format.duration ?? 0) * 1000);

      if (durationMs <= 0) {
        return resolve([]);
      }

      let stderrOutput = "";

      ffmpeg(filePath)
        .outputOptions([
          "-vf",
          `select='gt(scene\\,${threshold})',showinfo`,
          "-f",
          "null",
        ])
        .output("-")
        .on("stderr", (line: string) => {
          stderrOutput += line + "\n";
        })
        .on("end", () => {
          const lines = stderrOutput.split("\n");
          const showInfoRegex = /pts_time:(\d+\.?\d*)/;
          const sceneScoreRegex = /scene:(\d+\.?\d*)/;

          for (const line of lines) {
            if (!line.includes("showinfo") && !line.includes("pts_time")) continue;

            const timeMatch = showInfoRegex.exec(line);
            if (!timeMatch) continue;

            const timeSec = parseFloat(timeMatch[1]);
            let score = threshold;

            const scoreMatch = sceneScoreRegex.exec(line);
            if (scoreMatch) {
              score = parseFloat(scoreMatch[1]);
            }

            timestamps.push({ timeSec, score });
          }

          const scenes = buildScenes(timestamps, durationMs);
          resolve(scenes);
        })
        .on("error", (err) => {
          reject(new Error(`Scene detection failed: ${err.message}`));
        })
        .run();
    });
  });
}

/**
 * Converts a list of scene-change timestamps into scene segments.
 */
function buildScenes(
  timestamps: { timeSec: number; score: number }[],
  durationMs: number
): DetectedScene[] {
  if (timestamps.length === 0) {
    return [{ startMs: 0, endMs: durationMs, score: 0 }];
  }

  timestamps.sort((a, b) => a.timeSec - b.timeSec);

  const scenes: DetectedScene[] = [];
  let prevMs = 0;

  for (const ts of timestamps) {
    const currentMs = Math.round(ts.timeSec * 1000);
    if (currentMs > prevMs) {
      scenes.push({
        startMs: prevMs,
        endMs: currentMs,
        score: ts.score,
      });
    }
    prevMs = currentMs;
  }

  if (prevMs < durationMs) {
    scenes.push({
      startMs: prevMs,
      endMs: durationMs,
      score: 0,
    });
  }

  return scenes;
}

File: backend/app/services/video/scene_detector.py
"""Scene detection using ffmpeg — port of video/scene-detector.ts."""

import asyncio
import re


async def detect_scenes(file_path: str, threshold: float = 0.3) -> list[dict]:
    """
    Detect scene changes in a video using ffmpeg's scene filter.
    Returns a list of dicts with: start_ms, end_ms, score.
    """
    # First get the total duration
    probe_cmd = [
        "ffprobe", "-v", "quiet", "-show_entries", "format=duration",
        "-of", "csv=p=0", file_path,
    ]
    proc = await asyncio.create_subprocess_exec(
        *probe_cmd, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE,
    )
    stdout, _ = await proc.communicate()
    total_duration_s = float(stdout.decode().strip() or "0")
    total_duration_ms = int(total_duration_s * 1000)

    if total_duration_ms == 0:
        return []

    # Detect scene changes
    cmd = [
        "ffmpeg", "-i", file_path,
        "-vf", f"select='gt(scene,{threshold})',showinfo",
        "-f", "null", "-",
    ]

    proc = await asyncio.create_subprocess_exec(
        *cmd, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE,
    )
    _, stderr = await proc.communicate()

    output = stderr.decode()

    # Parse pts_time from showinfo output
    pattern = re.compile(r"pts_time:\s*([\d.]+)")
    timestamps_s = [float(m.group(1)) for m in pattern.finditer(output)]

    # Build scene boundaries
    scenes: list[dict] = []
    boundaries_ms = [0] + [int(t * 1000) for t in timestamps_s] + [total_duration_ms]

    for i in range(len(boundaries_ms) - 1):
        start = boundaries_ms[i]
        end = boundaries_ms[i + 1]
        if end > start:
            scenes.append({
                "start_ms": start,
                "end_ms": end,
                "score": threshold,
            })

    return scenes
